Update app-model documentation (#325)
napari/docs: docs/developers/architecture/app_model.md

# napari's application model

```{important}

> **This is not a part of the public napari interface!**

Reviewer: @lucyleeow I think it's probably still good to keep this notice here? Afaik there's no public interface for the app?

Author: No, but it is now inside the 'architecture' section which only includes docs aimed at devs, so we can infer that it is not part of the interface. I guess it's good to spell it out.

Reviewer 2: But as I understand, there is intention to make it public:
https://github.com/napari/napari/issues/6054

Without making get_app public, the actions cannot be simply called from plugins.

Author: If we do make it public we can amend this message.

I think the intention is more to allow users to register actions at runtime. Depending on the contribution type, the user may already have a public API func to access plugin actions- e.g., there is the `open_sample` func for sample contributions, and I think there may be a way for the user to get reader/writers too?

Reviewer 2: I think that option to call actions is as important, as runtime registration

Author: You think users should call a command in `app` via the command ID string instead of using existing public API like `open_sample`? Why?
Or do you just mean you want users to have the option of calling a command via string?

Reviewer 2: Because we provide public API only for opening sample and adding plugin widget. And plugins could provide many more contributions. 

And there is an area for people who would like to provide custom controllers (either widget or hardware)

Author: Sure but these are not the current comtribution types? We could create public API for new contribution types. Have we already decided how these new contributions would look?

I'm not opposed to execute via app model but there's probably a few decisions to be made here and this way may not be the best?

Reviewer: Yeah I don't think this is going to get hashed out in this PR lol.

Let's update the message to: This is not currently a part of the public napari API, but we are considering ways to expose action execution directly. 

Once we actually decide what the interface for that is, we will amend.

The napari application model is currently being developed. This document is here to
give guidance but things may change as we develop. It may link to current
pull requests and issues which have more up to date information on specific
areas.
```

## App-model

[`app-model`](https://app-model--142.org.readthedocs.build/en/142/) is a Python package
that provides a declarative schema for a GUI-based
application. It is an abstraction developed by napari developers, with the
needs of napari in mind, but it is agnostic to napari itself (i.e. it should be

> reusable by any python GUI application).

Reviewer: ```suggestion
reusable by any Python GUI application).
```

The {class}`~napari._app_model._app.NapariApplication` (`app`)
is the top level object that stores information about the commands, keybindings

> and menus that make up the appliation.

Reviewer: ```suggestion
and menus that make up the application.
```

It is a subclass of {class}`app_model.Application` and is a global application
singleton. It can be retrieved with {func}`napari._app_model.get_app`.

Currently, the primary purpose of the `app` is to compose the following
{mod}`app_model.registries` into a single name-spaced object:

* {class}`~app_model.registries.CommandsRegistry`: maintains all of the
  [commands](app-model-commands) (the actual callable objects) that have been
  registered with the application. Accessible via `app.commands`.
* {class}`~app_model.registries.MenusRegistry`: maintains all of the

>   [menu and submenus](appp-model-menus) that have been registered with the application.

Reviewer: ```suggestion
  [menus and submenus](appp-model-menus) that have been registered with the application.
```

  Accessible via `app.menus`.
* {class}`~app_model.registries.KeyBindingsRegistry`: maintains the association

>   between a [KeyBinding](app-model-keybindings) and a command id in the

Reviewer: ```suggestion
  between a [KeyBinding](app-model-keybindings) and a command ID in the
```

  {class}`~app_model.registries.CommandsRegistry`. Accessible via `app.keybindings`.

> The app-model

Reviewer: ```suggestion
The `app-model`
```

Author: Yeah I was not sure on this one. I think @jni said no backticks for napari, and I guess I thought it would be weird if napari package didn't get backticks but app-model did...? WDYT?

Reviewer 3: I think I'd prefer app-model in backticks to stand out more as something different than normal words.

['Getting started'](https://app-model--142.org.readthedocs.build/en/142/getting_started/)

> page provides a good general introduction to the app-model.

Reviewer: ```suggestion
page provides a good general introduction to the app model.
```

This documentation will focus on napari specific aspects.

(app-model-actions)=

## `Action`s

> The {class}`app_model.types.Action` class is designed to easily create high level

Reviewer: ```suggestion
The {class}`app_model.types.Action` class is designed to easily create a high level
```

'Action' object that is the "complete" representation of a command; a pointer
to a callable object and optionally placement in menus, keybindings, and additional
metadata like title, icons, tooltips etc. It subclasses
{class}`app_model.types.CommandRule` and takes {class}`app_model.types.MenuRule` and
{class}`app_model.types.KeyBindingRule` arguments.

The following code demonstrates the definition of a {class}`app_model.types.Action`
comprised of a "Split RGB" command, which is to be added to a specific section (group)
of the "layerlist context" menu, with a `Cmd+Alt+T` keybinding.

Note that while strings could be used for `id`, `title`, `menus.id` and
`keybindings.primary`, the usage of enums and constants makes refactoring and
maintenance much easier (and provides autocompletion in an IDE!). However, there is
currently intention to use plain strings for command `id` and `title` to simplify the

> decalaration of `Action`s.

Reviewer: ```suggestion
declaration of `Action`s.
```

```python
from app_model.types import Action, KeyMod, KeyCode
from napari._app_model.constants import CommandId, MenuId, MenuGroup
from napari._app_model.context import LayerListContextKeys as LLCK

> # `layers` will be injected layer when this action is invoked

Reviewer: ```suggestion
# `layers` will be injected later when this action is invoked
```

def split_rgb_layer(layers: 'LayerList'):
    ...

action = Action(
    id=CommandId.LAYER_SPLIT_RGB,
    title=CommandId.LAYER_SPLIT_RGB.title,
    callback=split_rgb_layer,
    menus = [
        {
            'id': MenuId.LAYERLIST_CONTEXT,
            'group': MenuGroup.LAYERLIST_CONTEXT.SPLIT_MERGE,
            'when': LLCK.active_layer_is_rgb,
        }
    ],
    keybindings=[{'primary': KeyMod.CtrlCmd | KeyMod.Alt | KeyCode.KeyT }]
)
```

`Action`s can be registered via {func}`app_model.Application.register_action`. This

> is essentially a shortcut for registering objects with the following registeries;

Reviewer: ```suggestion
is essentially a shortcut for registering objects with the following registries:
```

### `Action`s in napari

In napari non-Qt `Action`s are defined in
[`napari/_app_model/actions`](https://github.com/napari/napari/tree/main/napari/_app_model/actions)
while Qt `Action`s are defined in
[`napari/_qt/_qapp_model/qactions`](https://github.com/napari/napari/tree/main/napari/_qt/_qapp_model/qactions).
Non-Qt `Action`s get registered with `app` during
initialization of the napari `app`, in `NapariApplication`'s
{meth}`~napari._app_model._app.NapariApplication.__init__`. Qt `Action`s
get registered in
{func}`~napari._qt._qapp_model.qactions.init_qactions`, which gets
called during initialization of `_QtMainWindow`. Note this is relevant when
considering the differences between app-model and action manager implementation,
see [](app-model-action-manager-differences) for more.

(app-model-commands)=

## Commands

Commands represent callable objects that "do something" in napari, and usually
have a corresponding representation somewhere in the GUI (such as a button, a
menu item, or a keybinding).

Inline (lambda) and nested functions should be avoided for command callbacks to
prevent memory leakage.

All commands have a string id (e.g. '`napari:layer:duplicate`'). These are

> currently `CommandID` enums but we are currently considering changing to using

Reviewer: ```suggestion
currently `CommandID` enums but we are considering changing to using
```

plain strings to simplify decalaration of `Action`s.

{class}`app_model.types.CommandRule` class (of which {class}`app_model.types.Action`
is a subclass) includes many fields to detail various aspects of commands.
See {class}`app_model.types.CommandRule` for details on every field.
Notable {class}`~app_model.types.CommandRule` fields
are `enablement` and `toggled`, which both can take an
{class}`app_model.expressions.Expr`, an expression that can be evaluated, with
context if required (see [](context-expressions) for more details).

> `enablement` determines whether the command is enabled in the UI. It does not prevent

Reviewer: ```suggestion
`enablement` determines whether the command is visible in the UI. It does not prevent
```

Author: Nah `when` controls visibility. A visible, disabled action appears gray-ed out on the menu.

the command from being executed via other means, e.g., via
{meth}`app_model.registries.CommandsRegistry.execute_command`. `toggled` determines
whether the command appears checked/toggled in GUI representations (e.g., menu or
button). It can also take a {class}`app_model.types.ToggleRule`, which allows

> you to use a callable (instead of a expression) to determine toggle state.

Reviewer: ```suggestion
you to use a callable (instead of an expression) to determine toggle state.
```

### Commands should *not* be confused with the public napari API

While it is conceivable that plugins might need/want to refer to one of these
napari commands by its string id, it is **not** currently a goal that napari
end-users would execute any of these commands by their ID.  There should always
be a "pure python" way to import a napari object and call it. For example,
sample data is added to the `File` -> `Open Sample` menu but you can also
use {meth}`napari.Viewer.open_sample` to open a sample data in headless mode.

Commands mostly serve as a way to reference some functionality that needs to be exposed
in the GUI.

```{note}
Some of these command ID strings MAY be exposed externally in the future. For example, a plugin may wish to refer to a napari command.
```

### Commands in napari

> Commands are usually defined with their Action definition or with it's class,

Reviewer: ```suggestion
Commands are usually defined with their Action definition or with its class,
```

```{important}
**This is not a part of the public napari interface!**
```

```{warning}
**Work in progress!**
The napari application model is currently being developed. This document is here to
give guidance but things may change as we develop. It may link to current
pull requests and issues which have more up to date information on specific
areas.
```

## App-model

> [App-model](https://app-model--142.org.readthedocs.build/en/142/) is a Python package

Reviewer 3: This link gives me a 404, was there a reason it's that instead of the general one:
https://app-model.readthedocs.io/en/latest/

Reviewer 3: I'd put this mention of app-model in backticks too, see below.

Author: No reason, I think I've had the app model docs open in a tab for a long time and theres been a new release since and when I copied the URL it was to an older version? No idea, I will fix!

reusable by any Python GUI application).

The {class}`~napari._app_model._app.NapariApplication` (`app`)
is the top level object that stores information about the commands, keybindings
and menus that make up the application.
It is a subclass of {class}`app_model.Application` and is a global application
singleton. It can be retrieved with {func}`napari._app_model.get_app`.

Currently, the primary purpose of the `app` is to compose the following
{mod}`app_model.registries` into a single name-spaced object:

* {class}`~app_model.registries.CommandsRegistry`: maintains all of the
  [commands](app-model-commands) (the actual callable objects) that have been
  registered with the application. Accessible via `app.commands`.
* {class}`~app_model.registries.MenusRegistry`: maintains all of the
  [menus and submenus](appp-model-menus) that have been registered with the application.
  Accessible via `app.menus`.
* {class}`~app_model.registries.KeyBindingsRegistry`: maintains the association
  between a [KeyBinding](app-model-keybindings) and a command ID in the
  {class}`~app_model.registries.CommandsRegistry`. Accessible via `app.keybindings`.

The app-model

> ['Getting started'](https://app-model--142.org.readthedocs.build/en/142/getting_started/)

Reviewer 3: This link is also borked, see above.

page provides a good general introduction to app-model.
This documentation will focus on napari specific aspects.

(app-model-actions)=

## `Action`s

The {class}`app_model.types.Action` class is designed to easily create a high level

> 'Action' object that is the "complete" representation of a command; a pointer

Reviewer 3: ```suggestion
'Action' object that is the "complete" representation of a command: a pointer
```

        }
    ],
    keybindings=[{'primary': KeyMod.CtrlCmd | KeyMod.Alt | KeyCode.KeyT }]
)
```

`Action`s can be registered via {func}`app_model.Application.register_action`. This
is essentially a shortcut for registering objects with the following registries;
{class}`~app_model.registries.CommandsRegistry`,
{class}`~app_model.registries.MenusRegistry` and
{class}`~app_model.registries.KeyBindingsRegistry`. Note that a command ID may
currently only be registered once though this MAY change in the future if a need arises.

The code below shows how to register the `action` defined above with the napari
singleton `app`:

```python
from napari._app_model import get_app

get_app().register_action(action)
```

````{note}
If you're following along in the console, you may see the following error
when executing the above code:

```python
ValueError: Command 'napari:layer:split_rgb' already registered
```

This is because command id's may currently only be registered once, and associated with a single callback (and napari's internal app already used the `CommandId.LAYER_SPLIT_RGB` id).
````

(app-model-actions-napari)=

### `Action`s in napari

> In napari non-Qt `Action`s are defined in

Reviewer 3: ```suggestion
In napari, non-Qt `Action`s are defined in
```

[`napari/_app_model/actions`](https://github.com/napari/napari/tree/main/napari/_app_model/actions)
while Qt `Action`s are defined in
[`napari/_qt/_qapp_model/qactions`](https://github.com/napari/napari/tree/main/napari/_qt/_qapp_model/qactions).
Non-Qt `Action`s get registered with `app` during
initialization of the napari `app`, in `NapariApplication`'s
{meth}`~napari._app_model._app.NapariApplication.__init__`. Qt `Action`s
get registered in
{func}`~napari._qt._qapp_model.qactions.init_qactions`, which gets
called during initialization of `_QtMainWindow`. Note this is relevant when
considering the differences between app-model and action manager implementation,
see [](app-model-action-manager-differences) for more.

(app-model-commands)=

## Commands

Commands represent callable objects that "do something" in napari, and usually
have a corresponding representation somewhere in the GUI (such as a button, a
menu item, or a keybinding).

Inline (lambda) and nested functions should be avoided for command callbacks to

> prevent memory leakage.

Reviewer 3: Worth putting in an `important` admonition?

Author: Good idea!

]
```

To specify which menu(s) to add an Action to, use the `menus` field in
{class}`~app_model.types.Action`. `menus` takes a list of
{class}`~app_model.types.MenuRule`. The `id` field allows you to specify which menu
or submenu to add the Action to. All napari menus and submenus will also have a string
id (e.g. `'napari/file'`, or `'napari/layers/context'`), which should be declared as a
member of the {class}`~napari._app_model.constants.MenuId` enum. Internally, an instance
of this enum should be used instead of the string literal when referring to a menu.

The `group` and `order` fields let you control how menu items are grouped into sections,

> and they are ordered within each section. See {class}`~app_model.types.MenuItemBase` for

Reviewer: ```suggestion
and how they are ordered within each section. See {class}`~app_model.types.MenuItemBase` for
```

which will let you provide keybindings for all three operating systems.

### Keybindings in napari

App-model keybindings are currently only being used for menu bar items that
have migrated to using app-model `Action`s.
[Pull request 6204](https://github.com/napari/napari/pull/6204) aims to
migrate layer and viewer actions to app-model (though it is likely this pull request
will be split into several smaller pull requests for ease of review and better git
history). This pull request also implements [NAP-7](nap-7), which addresses
how keybindings are dispatched according to their priority, enablement, and potential
conflicts.

Currently keybindings are not able to be edited by the user at runtime but this
is planned. See [issue 6600](https://github.com/napari/napari/issues/6600) for more
details.

## Dependency injection and result processing

Dependency injection allows to write functions using parameter type annotations,
and then inject dependencies (arguments) into those functions at call time.
Very often in a GUI application, you may wish to infer some command arguments from
the current state of the application. For example, if you have menu item linked

> to a "close window", you likely want to close the *current* window.

Reviewer: ```suggestion
to a "close window" action, you likely want to close the *current* window.
```

Dependency injection enables this by providing arguments to commands at runtime, based

> on the type annotations in the command function definition. A 'provider', a function

Reviewer: ```suggestion
on the type annotations in the command function definition. A `provider`, a function
```

Author: hmm not sure about this one. When should we use `'` and when should we use backtick for provider/processor ?

Reviewer 3: I generally think of backticks for Python code, so function/method/variable/attribute names etc. (this is why I would put `app-model` and `in-and-out` in backticks, code-style)
" would then be used to indicate/emphasize a specific word being used in (or out of) a specific context vs. necessarily it's dictionary meaning. I'd avoid ' personally.

Author: Sorry to clarify, you would use `"provider"` instead of `'provider'` ?

Reviewer 3: If provider is not a Python "thing", I'd use "

Author: Done!

that can be called to return an instance of a given type, is used to obtain the
dependency to be injected.

Result processing allows you to process the return value of the command function at
execution time, based on command return type annotations.
For example, you may wish to automatically add layer data output from a command

> to the viewer. It is performed by 'processors', functions that accept an instance of a

Reviewer: ```suggestion
to the viewer. It is performed by `processors`, functions that accept an instance of a
```

given type and do something with it. Note that any value returned by a processor will
be ignored, it is the 'processor' function side effects that perform the desired
action (e.g., adding layer data to the viewer).

Dependency injection and result processing are provided by the package
[in-n-out](https://ino.readthedocs.io/en/latest/) (which spun out of an
internal napari module). App-model uses in-n-out to inject dependencies into all
commands in the {class}`~app_model.registries.CommandsRegistry`.
See the [in-n-out documentation](https://ino.readthedocs.io/en/latest/getting_started/)
for details on how providers and processors work.

In napari, providers and processors can be registered in the `app`'s
{attr}`~app_model.Application.injection_store` attribute, which is an instance of
[`in_n_out.Store`](https://ino.readthedocs.io/en/latest/reference/#in_n_out.Store).
This `Store` is just a collection of providers and processors. Providers

> and processors in the `Store` will be automatically be used when executing

Reviewer: ```suggestion
and processors in the `Store` will automatically be used when executing
```

napari to determine what providers it will make available, and what type hints
plugins/users may use to request dependencies.

### Annotation and namespaces

For parameter annotations to be resolved by in-n-out, the object must be in the
"global" or "local" namespace. The "global" namespace consists of the items

> in the callback functions {py:attr}`funciton.__globals__` attribute.

Reviewer: ```suggestion
in the callback function's {py:attr}`function.__globals__` attribute.
```

In in-n-out the "local" namespace consists of objects in the {py:mod}`typing`
and {py:mod}`types` modules' namespace, and objects in the
napari {attr}`~app_model.Application.injection_store`'s `namespace` attribute.

> In the napari `app`s `Store`, some basic napari objects are added to `namespace`

Reviewer: ```suggestion
In the napari `app`s `Store`, some basic napari objects are added to the `namespace`
```

attribute:

* all public types from {mod}`napari.components`, {mod}`napari.layers`
  and {mod}`napari.viewer`
* {class}`~napari._qt.qt_main_window.Window`
* {class}`~napari._qt.qt_viewer.QtViewer`

> This means that when annotating an callback function, you will not need to import

Reviewer: ```suggestion
This means that when annotating a callback function, you will not need to import
```

### Providers and processors in napari

Non-Qt providers and processors are defined in
[`napari/_app_model/injection`](https://github.com/napari/napari/tree/main/napari/_app_model/injection).
Qt providers and processors are defined in
[`napari/_qt/_qapp_model/injection`](https://github.com/napari/napari/tree/main/napari/_qt/_qapp_model/injection).

Non-Qt providers and processors are registered in the `app` `Store` during
initialization of the napari `app`, in `NapariApplication`'s
{meth}`~napari._app_model._app.NapariApplication.__init__`.
Qt providers and processors are registered in
{func}`~napari._qt._qapp_model.qactions.init_qactions`, which gets called during
initialization of `_QtMainWindow`. This is the same as
[registration of `Action`s](app-model-actions-napari).

## app-model testing

This section provides a guide to testing app-model aspects of napari. For general
information on napari testing see [](napari-testing).

### Mock app

The {class}`~napari._app_model._app.NapariApplication`, `app`, is a global application
singleton. This is not ideal because changes made in a previous test will persist and

> can cause conflicts for subsequent tests, when running a suit of tests.

Reviewer: ```suggestion
can cause conflicts for subsequent tests, when running a suite of tests.
```

Segmentation faults can also occur as a
singleton `app` may keep a reference to an object, e.g., a
{class}`~napari._qt.qt_main_window.Window`, that has
since been cleaned up at the end of a previous test.
Thus we mock the `app` and
[autouse](https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#autouse-fixtures-fixtures-you-don-t-have-to-request)

> it so a new instance of `app` is returned everytime {func}`napari._app_model.get_app`

Reviewer: ```suggestion
it so a new instance of `app` is returned every time {func}`napari._app_model.get_app`
```

> is used inside a test.

Reviewer: ```suggestion
is used inside a test. Note that a new `app` instance is also created when the `make_napari_viewer` fixture is used.
```

Author: `get_app` is called in `init_qactions` which is called in `make_napari_viewer` so yeah you're right. It is not returned though.

Author: So it seems that `_mock_app` is called during setup of any test (meaning a `app` is created for every test?). Is this because of the `autouse` ?

Author: (yes it is due to `autouse` that `_mock_app` fixture is called during setup of every test)

Author: I've added a note about this.

The mock `app` registers non-Qt `Action`s, providers and processors. This is
because it was thought that these would be required for the majority of tests.
If a plain `app` is required, it can be created by making a new
{class}`~napari._app_model._app.NapariApplication` instance.

> Qt items are not registered because we did not think it would not be best practice to

Reviewer: ```suggestion
Qt items are not registered because we did not think it would be best practice to
```

## Contexts

A context is essentially a mapping between variable names and their values.
For details on contexts and expressions see [](context-expressions).

There are currently two
{class}`~napari._app_model.context._context_keys.ContextNamespace` classes in
napari; {class}`~napari._app_model.context.LayerListContextKeys` and
{class}`~napari._app_model.context.LayerListSelectionContextKeys`. These map
variables to various values obtained using {class}`~napari.components.LayerList`
and selection {class}`~napari.components.LayerList` (e.g., "Number of selected points
layers"). They are defined in
[`napari/_app_model/context`](https://github.com/napari/napari/tree/main/napari/_app_model/context).
An instance of each class is created and saved as an attribute of
{class}`~napari.components.LayerList`. The
{meth}`~napari._app_model.context._context_keys.ContextNamespace.update`
methods of each are then connected to events that change the
{class}`~napari.components.LayerList` or selected {class}`~napari.components.LayerList`.

Variables in these classes can be used in expressions
in the [`Action`](app-model-actions) `enablement` field and

> menus {class}`~app_model.types.MenuRule`'s `when` field (see [](app-model-menus

Reviewer 3: ```suggestion
menus {class}`~app_model.types.MenuRule`'s `when` field (see [](app-model-menus)
```

1. It gives us an abstraction layer on top of Qt that will make it much easier to explore different application backends (such as a web-based app, etc..)
1. It's easier to test: `app-model` can take care of making sure that commands, menus, keybindings, and actions are rendered, updated, and triggered correctly, and napari can focus on testing the napari-specific logic.
1. It's becomes **much** easier to add & remove contributions from plugins if our internal representation of a command, menu, keybinding is similar to the schema that plugins use. The previous procedural approach made this marriage much more cumbersome.
1. **The Dream**: The unification of napari commands and plugin commands into a registry that can execute commands in response to user input provides an excellent base for "recording" a user workflow.  If all GUI user interactions go through dependency-injected commands, then it becomes much easier to export a script that reproduces a set of interactions.

(app-model-action-manager-differences)=

## app-model vs action manager implementation differences

App-model and action manager differ in when actions are executed.

> In app-model, actions are defined executed when they

Reviewer 3: > actions are defined executed when

This is confusing to me, either one of those should be dropped (executed?) or there is an "and" missing?

Author: Oh yeah I mucked that up. I think I mean 'executed'? (because define just means which file they are written in right??). When a function is connected to an event or registered to the `app`, the function itself is 'run' right? I mean we import the e.g., view menu actions and that import runs the functions inside the file where the view menu actions are defined?!

What is the best phrasing here?

cc @DragaDoncila

Reviewer: I would not use `executed` in this context. I would say "defined" when referring to the file they are literally written in, and `registered` when they are added to the app-model/action_manager registry.

Author: Yeah good point. I've remembered what I was thinking about when I wrote this -> plugins menu actions used to be registered later when we create the plugins menu (when we know we have Qt). Now we register plugin actions much earlier, when we don't know if we have Qt. (Thus we've had to do 'safe register')

Author: Done @DragaDoncila, let me know what you think, thanks!
